### PR TITLE
Allow proxy to be configurable for the timelock client

### LINKED ISSUE #2379: Allow proxy to be configurable for the timelock client.
We are trying to use atlasdb HA with timelock service. but we have a special setup where we set a proxy on the JVM (due to aws kinesis client). we would like to make a change to pass through proxy settings all the way down to the fiegn client here:(https://github.com/palantir/atlasdb/blob/9a126ef2654abcbbc7352a8d03f758f9bf5b7350/atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java#L100)

The proxy config we are looking at: https://github.com/palantir/http-remoting-api/blob/develop/service-config/src/main/java/com/palantir/remoting/api/config/service/ProxyConfiguration.java

@leeavital for SA.

### PR DESCRIPTION
**Pull request by amerguy**
Adding an optional ProxyConfiguration to TimelockConfigs to allow for custom proxy selector

**Goals (and why)**:
Fixes #2379. 

**Implementation Description (bullets)**:
Simply accept an optional ProxyConfiguration block that if set, will be used to get a proxySelector. which is then passed to the FeignOkHttpClients creator. 
The proxy selector logic would go away once we switch to using remoting for clients creation.
**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:
AtlasConfigs
**Priority (whenever / two weeks / yesterday)**:
Should be a small and harmless change that would unblock us. unless I am missing something.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2393)

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java [hunk 1]
```diff
@@ -84,33 +85,39 @@ private AtlasDbHttpClients() {
...
                 type,
                 UserAgents.DEFAULT_USER_AGENT);
     }
 
     public static <T> T createProxyWithFailover(
             Optional<SSLSocketFactory> sslSocketFactory,
+            Optional<ProxySelector> proxySelector,
             Collection<String> endpointUris,
             Class<T> type,
             String userAgent) {
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java]: The original methods are part of the public API and are used e.g. in internal Atlas coordination product, so we should leave them as they are (maybe mark them as @Deprecated) and have them call the new methods (with proxyselector) with an empty Optional.

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java [hunk 2]
```diff
@@ -89,17 +91,21 @@ private FeignOkHttpClients() {
...
             Optional<SSLSocketFactory> sslSocketFactory,
+            Optional<ProxySelector> proxySelector,
             String userAgent) {
-        return new OkHttpClient(newRawOkHttpClient(sslSocketFactory, userAgent));
+        return new OkHttpClient(newRawOkHttpClient(sslSocketFactory, proxySelector, userAgent));
     }
 
     @VisibleForTesting
-    static okhttp3.OkHttpClient newRawOkHttpClient(Optional<SSLSocketFactory> sslSocketFactory,
+    static okhttp3.OkHttpClient newRawOkHttpClient(
```
Comment (jeremyk-91) [inline atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java]: We need to change the `FeignOkHttpClientsTest` to supply an empty optional here too.

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java [hunk 3]
```diff
@@ -89,17 +91,21 @@ private FeignOkHttpClients() {
      */
     public static <T> Client newOkHttpClient(
             Optional<SSLSocketFactory> sslSocketFactory,
+            Optional<ProxySelector> proxySelector,
```
Comment (jeremyk-91) [inline atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java]: While technically a public API break this one might be okay, as no one uses it internally when I checked.

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java [hunk 4]
```diff
@@ -16,6 +16,8 @@
 package com.palantir.atlasdb.http;
 
 import java.io.IOException;
+import java.net.Proxy;
```
Comment (jeremyk-91) [inline atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java]: Probably need to remove this - it's unused (baseline will complain)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java (hunk 1 above)
Comment (amerguy) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java]: should the method name stay the same for the new methods ? (they will be overloaded with different parameter set)

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java (hunk 4 above)
Comment (amerguy) [reply atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java]: done

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java (hunk 2 above)
Comment (amerguy) [reply atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FeignOkHttpClients.java]: done

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java]: Yep, overloading is fine - we do this pretty often in Atlas

**Comment by jeremyk-91** [comment]
Note: Should be good to go after #2390 and/or if 2390 stalls.

**Comment by hsaraogi** [comment]
*[atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java, line 79 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTc4jH-6DeQt6Z3oJ:-KvSTc4jH-6DeQt6Z3oK:blaoacp) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java#L79)):*
> ```Java
> 
>     /**
>      * @deprecated please use createProxyWithFailover, which requires you to specify the ProxySelector parameter.
> ```

Please link to the alternative method we want users to use.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by hsaraogi** [comment]
*[atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java, line 95 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTejSpC-YQxggdBBB:-KvSTejSpC-YQxggdBBC:blaoacp) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java#L95)):*
> ```Java
> 
>     /**
>      * @deprecated please use createProxyWithFailover, which requires you to specify the ProxySelector parameter.
> ```

Please link to the alternative method we want users to use.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by hsaraogi** [comment]
*[atlasdb-config/src/test/java/com/palantir/atlasdb/http/AtlasDbHttpClientsTest.java, line 86 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTntFY25d6dhUt_rC:-KvSTntFY25d6dhUt_rD:bqp93nc) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/test/java/com/palantir/atlasdb/http/AtlasDbHttpClientsTest.java#L86)):*
> ```Java
> 
>         TestResource client = AtlasDbHttpClients.createProxyWithFailover(NO_SSL,
>                 Optional.empty(), bothUris, TestResource.class);
> ```

Add a test with DIRECT amd HTTP proxy selectors here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by hsaraogi** [comment]
*[docs/source/release_notes/release-notes.rst, line 48 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSVFTX8Mxs8YwhiLY2:-KvSVFTX8Mxs8YwhiLY3:b-1595po) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/docs/source/release_notes/release-notes.rst#L48)):*
> ```reStructuredText
> 
>     *    - |improved|
>          - Allow passing a proxyConfiguration to allow setting custom proxy on the TimeLock clients.
> ```

Technically, you can set this up on lock and timestamp blocks too. Should that be a part of the release notes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by hsaraogi** [comment]
Reviewed 5 of 25 files at r1, 2 of 3 files at r2, 3 of 3 files at r4, 16 of 17 files at r5.
Review status: 26 of 27 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by gsheasby** [comment]
Review status: 26 of 27 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java, line 79 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTc4jH-6DeQt6Z3oJ:-KvSqeD7oiYtWI0OP2K6:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java#L79)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Please link to the alternative method we want users to use.
</blockquote></details>

Done

---

*[atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java, line 95 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTejSpC-YQxggdBBB:-KvSqenr7tXw8XYzftBn:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java#L95)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Please link to the alternative method we want users to use.
</blockquote></details>

Done

---

*[docs/source/release_notes/release-notes.rst, line 48 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSVFTX8Mxs8YwhiLY2:-KvSqa7Q_gshaSAJTttw:b-kq1ovt) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/docs/source/release_notes/release-notes.rst#L48)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Technically, you can set this up on lock and timestamp blocks too. Should that be a part of the release notes?
</blockquote></details>

Agreed that some form of documentation should be linked to here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by amerguy** [comment]
Review status: 24 of 27 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[atlasdb-config/src/test/java/com/palantir/atlasdb/http/AtlasDbHttpClientsTest.java, line 86 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTntFY25d6dhUt_rC:-KvTmvXP9ALpo-Lo7o7R:bh7y9al) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/test/java/com/palantir/atlasdb/http/AtlasDbHttpClientsTest.java#L86)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary><blockquote>

Add a test with DIRECT amd HTTP proxy selectors here.
</blockquote></details>

What is the test actually testing ? the only thing we are doing here is creating a proxySelector from a proxyConfig. This code already exist and tested in http-remoting (and should be used out of the box once we migrate to remoting 3).
The other thing we could be testing is that setting `proxySelector` property on an `OkHttpClient`actually sets the proxySelector. if you can be more specific on what the tests would actually check am happy to add it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by amerguy** [comment]
Review status: 24 of 27 files reviewed at latest revision, 8 unresolved discussions.

---

*[docs/source/release_notes/release-notes.rst, line 48 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSVFTX8Mxs8YwhiLY2:-KvXzwqe-lhtlgniwgJQ:b-24q3ea) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/docs/source/release_notes/release-notes.rst#L48)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Agreed that some form of documentation should be linked to here.
</blockquote></details>

linked to proxyConfig docs

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by hsaraogi** [comment]
*[atlasdb-config/src/test/java/com/palantir/atlasdb/http/AtlasDbHttpClientsTest.java, line 86 at r5](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvSTntFY25d6dhUt_rC:-KvbQPfrdUaAUddcePZI:b3zvufh) ([raw file](https://github.com/palantir/atlasdb/blob/25c422cd30c7ae50687185d38ff12d465af4fc1a/atlasdb-config/src/test/java/com/palantir/atlasdb/http/AtlasDbHttpClientsTest.java#L86)):*
<details><summary><i>Previously, amerguy wrote…</i></summary><blockquote>

What is the test actually testing ? the only thing we are doing here is creating a proxySelector from a proxyConfig. This code already exist and tested in http-remoting (and should be used out of the box once we migrate to remoting 3).
The other thing we could be testing is that setting `proxySelector` property on an `OkHttpClient`actually sets the proxySelector. if you can be more specific on what the tests would actually check am happy to add it.

</blockquote></details>

We are testing that the proxy is properly setup on the Atlas client here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by hsaraogi** [comment]
Reviewed 1 of 17 files at r5.
Review status: 24 of 27 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[docs/source/release_notes/release-notes.rst, line 51 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/2393#-KvbR1cfL5AHTPCm6JDY:-KvbR1cfL5AHTPCm6JDZ:b-ya3h17) ([raw file](https://github.com/palantir/atlasdb/blob/a30b472f23cf01cd68d1883cef2fc96b214f0d71/docs/source/release_notes/release-notes.rst#L51)):*
> Agreed that some form of documentation should be linked to here.
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>
>               
>               
>                 
>                   1 day ago
>                 
>               
>             
>               
>               
>               amerguy
</details>
> 
>               
>               
>                 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2393)*

**Comment by fsamuel-bs** [comment]
@amerguy looks like one of your tests is failing with a 404:

```
com.palantir.atlasdb.http.AtlasDbHttpClientsTest > httpProxyIsConfigurableOnClientRequests FAILED
java.lang.RuntimeException: Error 404. Reason: Not Found. Body:
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"/>
<title>Error 404 NOT_FOUND</title>
</head>
<body><h2>HTTP ERROR 404</h2>
<p>Problem accessing /__files/number. Reason:
<pre>    NOT_FOUND</pre></p><hr /><i><small>Powered by Jetty://</small></i><br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                
<br/>                                                

</body>
</html>


	at com.palantir.remoting2.errors.SerializableErrorToExceptionConverter.getException(SerializableErrorToExceptionConverter.java:67)
	at com.palantir.atlasdb.http.SerializableErrorDecoder.decode(SerializableErrorDecoder.java:42)
	at com.palantir.atlasdb.http.AtlasDbErrorDecoder.decode(AtlasDbErrorDecoder.java:42)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:134)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:76)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:103)
	at com.sun.proxy.$Proxy13.getTestNumber(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.palantir.tritium.proxy.InvocationEventProxy.execute(InvocationEventProxy.java:138)
	at com.palantir.tritium.proxy.InvocationEventProxy.instrumentInvocation(InvocationEventProxy.java:115)
	at com.palantir.tritium.proxy.InvocationEventProxy.handleInvocation(InvocationEventProxy.java:95)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87)
	at com.sun.proxy.$Proxy13.getTestNumber(Unknown Source)
	at com.palantir.atlasdb.http.AtlasDbHttpClientsTest.httpProxyIsConfigurableOnClientRequests(AtlasDbHttpClientsTest.java:132)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:62)
	at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:62)
	at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:62)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

I was able to reproduce this locally as well. We're going ahead with the 0.59.0 release without this change, but when you've fixed this error, ping us and we can release again with it. :)

**Pull request merged**